### PR TITLE
add validation in all existing apis

### LINKED ISSUE #3810: Implement validation across all existing apis
Parent issue: #3765 
Implement validation from http://marshmallow.readthedocs.io/en/latest/quickstart.html#validation

### PR DESCRIPTION
**Pull request by shubham-padia**
also merge two migration heads

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

#### Changes proposed in this pull request:

- add validation in all existing apis
- also merge two migration heads
- move save_to_db from api/helpers/helpers.py to api/helpers/db.py and delete api/helpers/helpers.py
- Unprocessable entity: 422 has been used for validation errors

Ref:
- http://marshmallow.readthedocs.io/en/latest/quickstart.html#validation
- http://marshmallow.readthedocs.io/en/latest/extending.html#schema-level-validation
- http://marshmallow.readthedocs.io/en/latest/_modules/marshmallow/validate.html
- http://marshmallow.readthedocs.io/en/latest/api_reference.html#marshmallow.Schema.validate

Fixes #3810

### CODE REVIEW

**Comment by shubham-padia** [comment]
@SaptakS @niranjan94 @poush please review :)

File: app/api/event_copyright.py [hunk 1]
```diff
@@ -22,7 +22,7 @@ class Meta:
     holder_url = fields.Url()
     licence = fields.Str()
     licence_url = fields.Url()
-    year = fields.Int()
+    year = fields.Int(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/event_copyright.py]: And can we go for a sane year here ? Above 1900 and equal and below current.

File: app/api/discount_codes.py [hunk 2]
```diff
@@ -20,18 +23,23 @@ class Meta:
...
     value = fields.Float()
-    type = fields.Str()
+    type = fields.Str(validate=validate.OneOf(choices=["amount", "percent"]))
     is_active = fields.Boolean()
-    tickets_number = fields.Integer()
-    min_quantity = fields.Integer()
-    max_quantity = fields.Integer()
+    tickets_number = fields.Integer(validate=lambda n: n >= 0)
+    min_quantity = fields.Integer(validate=lambda n: n >= 0)
+    max_quantity = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/discount_codes.py]: Is it possible to check and ensure this is higher (or) equal to the `min_quantity` and at the same time less than or equal to `tickets_number` ?

File: app/api/discount_codes.py [hunk 3]
```diff
@@ -20,18 +23,23 @@ class Meta:
...
     discount_url = fields.Url()
     value = fields.Float()
-    type = fields.Str()
+    type = fields.Str(validate=validate.OneOf(choices=["amount", "percent"]))
     is_active = fields.Boolean()
-    tickets_number = fields.Integer()
-    min_quantity = fields.Integer()
-    max_quantity = fields.Integer()
+    tickets_number = fields.Integer(validate=lambda n: n >= 0)
+    min_quantity = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/discount_codes.py]: Is it possible to check and ensure this is lower (or) equal to the `max_quantity`  ?

File: app/api/event_invoices.py [hunk 4]
```diff
@@ -20,24 +22,25 @@ class Meta:
...
     completed_at = fields.DateTime(default=None)
     transaction_id = fields.Str()
-    paid_via = fields.Str()
+    paid_via = fields.Str(validate=validate.OneOf(choices=["free", "stripe", "paypal", "transfer", "onsite", "cheque"]))
     payment_mode = fields.Str()
     brand = fields.Str()
-    exp_month = fields.Integer()
-    exp_year = fields.Integer()
+    exp_month = fields.Integer(validate=lambda n: 0 <= n <= 12)
+    exp_year = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/event_invoices.py]: Can we go for a sane year here ? Above `2015`

File: app/api/helpers/static.py [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+        "Baseball", "Basketball", "Cycling", "Exercise", "Fighting & Martial Arts",
+        "Football", "Golf", "Hockey", "Motorsports", "Mountain Biking",
+        "Obstacles", "Other", "Rugby", "Running", "Snow Sports", "Soccer",
+        "Swimming & Water Sports", "Tennis", "Volleyball", "Walking", "Yoga"
+    ],
+    'Travel & Outdoor': [
+        "Canoeing", "Climbing", "Hiking", "Kayaking", "Other", "Rafting", "Travel"
+    ]
+}
+PAYMENT_COUNTRIES = [
```
Comment (niranjan94) [inline app/api/helpers/static.py]: Countries to be referred to by the ISO two letter country code only. 
https://github.com/fossasia/open-event-frontend/blob/development/app/utils/dictionary/demography.js

File: app/api/image_sizes.py [hunk 6]
```diff
@@ -23,20 +23,20 @@ class Meta:
...
-    icon_quality = fields.Integer()
-    thumbnail_width = fields.Integer()
-    thumbnail_height = fields.Integer()
+    full_width = fields.Integer(validate=lambda n: n >= 0)
+    full_height = fields.Integer(validate=lambda n: n >= 0)
+    full_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
+    full_quality = fields.Integer(validate=lambda n: n >= 0)
+    icon_width = fields.Integer(validate=lambda n: n >= 0)
+    icon_height = fields.Integer(validate=lambda n: n >= 0)
+    icon_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
```
Comment (niranjan94) [inline app/api/image_sizes.py]: This is boolean. Why the validation lambda ?

File: app/api/image_sizes.py [hunk 7]
```diff
@@ -23,20 +23,20 @@ class Meta:
...
-    full_quality = fields.Integer()
-    icon_width = fields.Integer()
-    icon_height = fields.Integer()
-    icon_aspect = fields.Boolean(default=False)
-    icon_quality = fields.Integer()
-    thumbnail_width = fields.Integer()
-    thumbnail_height = fields.Integer()
+    full_width = fields.Integer(validate=lambda n: n >= 0)
+    full_height = fields.Integer(validate=lambda n: n >= 0)
+    full_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
```
Comment (niranjan94) [inline app/api/image_sizes.py]: This is boolean. Why the validation lambda ?

File: app/api/image_sizes.py [hunk 8]
```diff
@@ -23,20 +23,20 @@ class Meta:
...
+    icon_height = fields.Integer(validate=lambda n: n >= 0)
+    icon_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
+    icon_quality = fields.Integer(validate=lambda n: n >= 0)
+    thumbnail_width = fields.Integer(validate=lambda n: n >= 0)
+    thumbnail_height = fields.Integer(validate=lambda n: n >= 0)
     thumbnail_aspect = fields.Boolean(default=False)
-    thumbnail_quality = fields.Integer()
-    logo_width = fields.Integer()
-    logo_height = fields.Integer()
+    thumbnail_quality = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/image_sizes.py]: And less than/equal to 100

File: app/api/image_sizes.py [hunk 9]
```diff
@@ -23,20 +23,20 @@ class Meta:
...
-    thumbnail_width = fields.Integer()
-    thumbnail_height = fields.Integer()
+    full_width = fields.Integer(validate=lambda n: n >= 0)
+    full_height = fields.Integer(validate=lambda n: n >= 0)
+    full_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
+    full_quality = fields.Integer(validate=lambda n: n >= 0)
+    icon_width = fields.Integer(validate=lambda n: n >= 0)
+    icon_height = fields.Integer(validate=lambda n: n >= 0)
+    icon_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
+    icon_quality = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/image_sizes.py]: And less than/equal to 100

File: app/api/image_sizes.py [hunk 10]
```diff
@@ -23,20 +23,20 @@ class Meta:
...
-    icon_width = fields.Integer()
-    icon_height = fields.Integer()
-    icon_aspect = fields.Boolean(default=False)
-    icon_quality = fields.Integer()
-    thumbnail_width = fields.Integer()
-    thumbnail_height = fields.Integer()
+    full_width = fields.Integer(validate=lambda n: n >= 0)
+    full_height = fields.Integer(validate=lambda n: n >= 0)
+    full_aspect = fields.Boolean(validate=lambda n: n >= 0, default=False)
+    full_quality = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/image_sizes.py]: And less than/equal to 100

File: app/api/session_types.py [hunk 11]
```diff
@@ -25,9 +28,16 @@ class Meta:
...
+    def validate_length(self, data):
+        try:
+            datetime.strptime(data['length'], '%H:%M')
+        except ValueError:
+            raise UnprocessableEntity({'pointer': 'length'}, "Length should be in the format %H:%M")
+
     id = fields.Str(dump_only=True)
     name = fields.Str(required=True)
-    length = fields.Str()
+    length = fields.Str(required=True)
```
Comment (niranjan94) [inline app/api/session_types.py]: In the format `hh:mm`

File: app/api/sessions.py [hunk 12]
```diff
@@ -40,9 +48,9 @@ class Meta:
     videos_url = fields.Url(attribute='videos')
     audios_url = fields.Url(attribute='audios')
     signup_url = fields.Url()
-    state = fields.Str()
-    created_at = fields.DateTime()
-    deleted_at = fields.DateTime()
+    state = fields.Str(validate=validate.OneOf(choices=["pending", "accepted", "confirmed", "rejected"]))
```
Comment (niranjan94) [inline app/api/sessions.py]: Add state `draft`

File: app/api/tax.py [hunk 13]
```diff
@@ -20,7 +20,7 @@ class Meta:
     id = fields.Str(dump_only=True)
     country = fields.Str()
     name = fields.Str(required=True)
-    rate = fields.Float(required=True)
+    rate = fields.Float(validate=lambda n: n >= 0, required=True)
```
Comment (niranjan94) [inline app/api/tax.py]: And less than equal to 100

File: app/api/tickets.py [hunk 14]
```diff
@@ -16,20 +18,25 @@ class Meta:
...
     position = fields.Integer()
     is_fee_absorbed = fields.Boolean()
-    sales_starts_at = fields.DateTime()
-    sales_ends_at = fields.DateTime()
+    sales_starts_at = fields.DateTime(required=True)
+    sales_ends_at = fields.DateTime(required=True)
     is_hidden = fields.Boolean(default=False)
-    min_order = fields.Integer()
-    max_order = fields.Integer()
+    min_order = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/tickets.py]: Is it possible to check and ensure this is lower (or) equal to the `max_order` ?

File: app/api/tickets.py [hunk 15]
```diff
@@ -16,20 +18,25 @@ class Meta:
...
+    max_order = fields.Integer(validate=lambda n: n >= 0)
```
Comment (niranjan94) [inline app/api/tickets.py]: Is it possible to check and ensure this is higher (or) equal to the `min_order` and at the same time less than or equal to `quantity` ?

File: app/api/discount_codes.py (hunk 3 above)
Comment (shubham-padia) [inline app/api/discount_codes.py]: already checking that at line 26 :)

File: app/api/discount_codes.py (hunk 2 above)
Comment (shubham-padia) [inline app/api/discount_codes.py]: Yes, it's possible, but I didn't quite get the logic of below:

> max_quantity less than or equal to tickets_number

File: app/api/session_types.py (hunk 11 above)
Comment (shubham-padia) [inline app/api/session_types.py]: already being checked at line 31 above ;)

File: app/api/tickets.py (hunk 15 above)
Comment (shubham-padia) [inline app/api/tickets.py]: You're talking about min_order <= quantity or max_order <= quantity ?

File: app/api/discount_codes.py (hunk 2 above)
Comment (niranjan94) [inline app/api/discount_codes.py]: `min_quantity <= max_quantity <= tickets_number`

File: app/api/tickets.py (hunk 15 above)
Comment (niranjan94) [inline app/api/tickets.py]: `min_order <= max_order <= quantity`

**Comment by shubham-padia** [comment]
@niranjan94 Sorry for the late reply. I was waiting for reviews since when and I missed this in the multiple notifications :) .
- Requested changes have been made
- Left comment at some places, please see them and review further.

File: app/api/discount_codes.py (hunk 2 above)
Comment (shubham-padia) [inline app/api/discount_codes.py]: I think i'm missing something here, Shouldn't it be :
`min_quantity <= tickets_number <= max_quantity `

Comment (niranjan94) [inline app/api/discount_codes.py]: Wait ... what is tickets number here ? 🤔

Comment (shubham-padia) [inline app/api/discount_codes.py]: okay. currently Tickets_number is the max no. of tickets to  be discounted and min and max quantity are min and max discounts per order.

Comment (niranjan94) [inline app/api/discount_codes.py]: Ohh .... okay ... I probably mis-understood. Then this should just be 
`min_quantity <= max_quantity`

Comment (shubham-padia) [inline app/api/discount_codes.py]: but your logic is correct,  `min_quantity <= max_quantity <= tickets_number`, I'll make the changes

**Comment by shubham-padia** [comment]
@niranjan94 all changes resolved

**Comment by shubham-padia** [comment]
@niranjan94 please review :)

File: app/api/helpers/static.py (hunk 5 above)
Comment (niranjan94) [inline app/api/helpers/static.py]: Now ... How does this switch affect existing data ?

File: app/api/events.py [hunk 16]
```diff
@@ -33,16 +35,36 @@ class Meta:
         self_view_many = 'v1.event_list'
         inflect = dasherize
 
+    @validates_schema
+    def validate_date(self, data):
+        if data['starts_at'] >= data['ends_at']:
+            raise UnprocessableEntity({'pointer': 'ends_at'}, "ends_at should be after starts_at")
```
Comment (niranjan94) [inline app/api/events.py]: Could you give me a sample of the error response ? 

And also, what's the difference b/w `pointer` and `parameter` ... Both seem to be used in diff places here.

Comment (shubham-padia) [inline app/api/events.py]:
```
{
    "errors": [
        {
            "status": 422,
            "source": {
                "pointer": "ends_at"
            },
            "detail": "ends_at should be after starts_at",
            "title": "Unprocessable Entity"
        }
    ],
    "jsonapi": {
        "version": "1.0"
    }
}
```
About pointer and parameter, we've been using pointer, parameter and source inside `source` in the error message interchangeably andwe should decide on one of them, what do you think?

Comment (shubham-padia) [inline app/api/events.py]: IMO `pointer` seems the best as it is regularly used in examples in http://jsonapi.org/examples/#error-objects

Comment (niranjan94) [inline app/api/events.py]: We'll go for pointer. Also, the attribute should be path and not just a name. `/data/attributes/end-at`

Ref: http://jsonapi.org/examples/#error-objects-basics

Comment (niranjan94) [inline app/api/events.py]: And also note that it is dasherized.

**Comment by shubham-padia** [comment]
> Now ... How does this switch affect existing data ?

Breaks backwards compatibility right now.
Data migrations can be provided for this. my event-invoices table data is completely blank right now but seeing from other country fields, there are many invalid country names with spelling mistakes and they can be lost, no alternative for them. Shouldn't we implement this switch for all country fields in a seperate issue, you can assign that to me ?

**Comment by niranjan94** [comment]
Yep. Cool ... will open new issue and assign to you

File: app/api/events.py (hunk 16 above)
Comment (shubham-padia) [inline app/api/events.py]: Okay, parameter is for query parameters. And pointer is for data attributes.

Comment (shubham-padia) [inline app/api/events.py]: Resolved.

**Comment by shubham-padia** [comment]
@niranjan94 please review :)

File: app/api/events.py [hunk 17]
```diff
@@ -33,16 +35,36 @@ class Meta:
...
+
+    @validates_schema
+    def validate_timezone(self, data):
+        if 'timezone' in data['timezone']:
```
Comment (LuD1161) [inline app/api/events.py]: @shubham-padia 
should have been `if 'timezone' in data`
Done changes in [PR#3857](https://github.com/fossasia/open-event-orga-server/pull/3857)

File: app/api/events.py [hunk 18]
```diff
@@ -33,16 +35,36 @@ class Meta:
...
+            offset = timezone(data['timezone']).tz.utcoffset(datetime.strptime('2014-12-01', '%Y-%m-%d')).seconds
```
Comment (LuD1161) [inline app/api/events.py]: @shubham-padia 
**tz** should be removed only 
`offset = timezone(data['timezone']).utcoffset(datetime.strptime('2014-12-01', '%Y-%m-%d')).seconds`
Done changes in [PR#3857](https://github.com/fossasia/open-event-orga-server/pull/3857)

**Pull request merged**